Creating a caching key for queries the H6 kind of way

ahus1 opened the pull request:
Closes #16332

<!---
Please read https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

ahus1 committed:
Creating a caching key for queries the H6 kind of way

Closes #16332

ahus1 commented:
@hmlnarik - this is now ready for review. A manual local test showed that this work now better when showing users related to a role. 

This only compiles and runs the unit tests, everything else is expected to fail until more of the Arquillian setup has been fixed.

vramik commented on model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapKeycloakTransaction.java:
```diff
@@ -171,13 +175,29 @@ public Stream<E> read(QueryParameters<M> queryParameters) {
+        // Add aliases for each root to avoid generating IDs with random object IDs in Hibernate 6.
+        // Those random object IDs would prevent creating a stable HQL string which is the key for caching.
+        // Use a HashMap to create increasing suffix numbers for each entity type.
+        Map<String, Integer> aliases = new HashMap<>();
+        for (Root<?> queryRoot : query.getRoots()) {
+            if (queryRoot.getAlias() == null) {
+                String name = queryRoot.getModel().getName();
+                Integer count = aliases.get(name);
+                if (count == null) {
+                    count = 0;
+                }
+                queryRoot.alias(name + count);
+                count ++;
+                aliases.put(name, count);
+            }
+        }
```
Thank you @ahus1!

It seems to me that for each call of `read(QueryParameters<M> queryParameters)` there is empty map (`aliases`) created and for each root, which doesn't have an alias, it's created one. Then the alias is inserted into `aliases`.

Could you please provide an example of a query with multiple roots where alias would be found on line 185? So far I was not able to came across such a query. Thank you

ahus1 replied:
@vramik - I'm not aware of a secondary root being used at the moment. The closes I found was in JpaRoleModelCriteriaBuilder the subquery that's being created. 

I tried to trigger this code by calling a delete on  a role, but this this failed earlier as the "->" JSON functions weren't working (that's possibly another issue that we're about to soon for the H6 upgrade).

So to summarize: At the moment it is used only for one root, and I am not sure what it would do on the subquery. If the code would not set an alias for whatever reason, it would lead to a query that can't be cached, which might turn up in a performance test eventually.  

This whole caching code is currently used to cache the realm when working via the admin UI (and REST endpoints). It is also used to cache a client if a recall correctly during refresh of a token. I verified that those caching works as expected.

It still has its justification IMHO, at the same time we might skip it if we see it adds to much complexity.

vramik replied:
Makes sense, thanks. As far as I could see caching works as expected. My understating of the code above was that aliases are only stored into a map but there is never alias found in line 185. 

I just wanted to understand the purpose of the map properly so I wonder whether my understanding is correct or not.

ahus1 replied:
@vramik - yes, your understanding is correct. As there is currently only a single root, the logic isn't really tested used yet.  Today all aliases have the "0" at the end. For that to be "1", it would need to have a root with the same name a second time. 

I was assigning a separate numbering for each root as for two roots, I wouldn't know in which order I would receive them from the set of roots. So it was better to number those wich the same name consecutively instead of numbering all aliases consecutively.

Let me know if this comment helped to get this PR towards an approval for a review, or if you'd prefer to have a short call.

vramik replied:
Thanks for the comment, makes sense to me. Approving.

ahus1 commented:
@hmlnarik - thanks for looking at the model tests, I missed that those should already run with H6. The error messages were suspicious, and I updated my code. 

The way to extract the values and and to generated artificial parameter names was incomplete and fragile. I now look that the value is part of the query string. 

This now fixed the tests when I ran them locally, CI is still running. 

Please have a look. 

This is all built on the assumption that if there is no query parameter name, Hibernate will have put it into the query string - and this is a very internal Hibernate behavior. And HQL strings are not well supported in H6 anyway, as they're moving towards JPQL even internally AFAIK.

I think looking into the query string is still too fragile and error prone, I'm ready to rip out all of it. It will affect the performance of the admin UI with regards to caching of the realm entity, but with other optimizations I made it is not as important as it was earlier. If we rip it out, we'll have to wait for #13641.

Switching this to a draft state due to this.

ahus1 commented:
@hmlnarik / @vramik - after talking to @hmlnarik and discussing my concerns with the solution that was dependent on  internal Hibernate workings, this PR is now updated to use the `ModelCriterialBuilder` directly to prepare the query. 

As this takes out the ugly bits, this is now ready for review again. Thanks!

hmlnarik commented on model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapKeycloakTransaction.java:
```diff
@@ -348,28 +345,19 @@ private Predicate notExpired(final CriteriaBuilder cb, final JpaSubqueryProvider
-            }
-            this.queryMaxResults = emQuery.getMaxResults();
-            this.queryFirstResult = emQuery.getFirstResult();
+            DefaultModelCriteria dmc = query.getModelCriteriaBuilder()
```
Nit: this could be checked to be `DefaultModelCriteria` already, in which case flashing could be skipped.

hmlnarik replied:
I agree

hmlnarik reviewed (commented):
Thank you @ahus1! The changes are so good that I'd propose opening them against `main` branch instead, WDYT?

ahus1 commented:
@hmlnarik - I'd like to merge this to the Quarkus 3 branch first as I want to see the model tests succeed with our three PRs. 

Once it is merged to Quarkus 3, I'll prepare a another PR for main. That would then lead to some merge conflicts on rebase for the Quarkus 3 branch.

vramik commented on model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapKeycloakTransaction.java:
```diff
@@ -174,10 +170,10 @@ public Stream<E> read(QueryParameters<M> queryParameters) {
 
         Map<QueryCacheKey, List<RE>> cache = getQueryCache();
-        QueryCacheKey queryCacheKey = new QueryCacheKey(emQuery, modelType);
+        QueryCacheKey queryCacheKey = null;
+        queryCacheKey = new QueryCacheKey(queryParameters, modelType);
```
Thank you @ahus1, I like the way it's done. Great idea.

I have just one nitpick: there is `null` assigned into `queryCacheKey` on line 173 and on 174 there is actual `QueryCacheKey`. Would it make sense:
```suggestion
        QueryCacheKey queryCacheKey = new QueryCacheKey(queryParameters, modelType);
```
?

ahus1 replied:
Will do. As Hynek suggested, I'm now preparing a new PR against main, see #16516 (it is currently in draft)
